io/private/las/Vlr.cpp

```diff
@@ -0,0 +1,412 @@
+
+    in >> recordSig;
+    in.get(userId, 16);
+    in >> recordId >> promisedDataSize;
+    in.get(description, 32);
+}
+
+std::vector<char> Evlr::headerData() const
+{
+    std::vector<char> buf(HeaderSize);
+
+    LeInserter out(buf.data(), HeaderSize);
+    out << recordSig;
+    out.put(userId, 16);
+    out << recordId << (uint64_t)dataVec.size();
+    out.put(description, 32);
+
+    return buf;
+}
+
+std::istream& operator>>(std::istream& in, las::Evlr& v)
+{
+    NL::json j;
+    in >> j;
+
+    // Make sure there isn't stuff in the input stream after the JSON
+    // object.
+    char c;
+    do
+    {
+        in >> c;
+    } while (in.good() && std::isspace(c));
+    if (!in.eof())
+        throw pdal_error("Invalid characters following LAS VLR JSON object.");
+
+    // We forced an EOF above, so clear the error state.
+    in.clear();
+    if (!j.is_object())
+        throw pdal_error("LAS VLR must be specified as a JSON object.");
+
+    std::string description;
+    std::string b64data;
+    std::string userId;
+    std::vector<char> data;
+    double recordId(std::numeric_limits<double>::quiet_NaN());
+    for (auto& el : j.items())
+    {
+        if (el.key() == "description")
+        {
+            if (!el.value().is_string()) 
+                throw pdal_error("LAS VLR description must be specified "
+                    "as a string.");
+            description = el.value().get<std::string>();
+            if (description.size() > 32)
+                throw pdal_error("LAS VLR description must be 32 characters "
+                    "or less.");
+        }
+        else if (el.key() == "record_id")
+        {
+            if (!el.value().is_number())
+                throw pdal_error("LAS VLR record ID must be specified as "
+                    "a number.");
+            recordId = el.value().get<double>();
+            if (recordId < 0 ||
+                recordId > (std::numeric_limits<uint16_t>::max)() ||
+                recordId != (uint16_t)recordId)
+                throw pdal_error("LAS VLR record ID must be an non-negative "
+                    "integer less than 65536.");
+        }
+        else if (el.key() == "user_id")
+        {
+            if (!el.value().is_string()) 
+                throw pdal_error("LAS VLR user ID must be specified "
+                    "as a string.");
+            userId = el.value().get<std::string>();
+            if (userId.size() > 16)
+                throw pdal_error("LAS VLR user ID must be 16 characters "
+                    "or less.");
+        }
+        else if (el.key() == "data")
+        {
+            if (data.size())
+                throw pdal_error("Can't specify both 'data' and 'filename' "
+                    "in VLR specification.");
+            if (!el.value().is_string())
+                throw pdal_error("LAS VLR data must be specified as "
+                    "a base64-encoded string.");
+            //ABELL - Fix this.
```

Comment:
convert to using `char` too?

Reply:
It's code that people might be using. Didn't want to break.